Enhancement: Add modDirectory getList() method to modFileHandler

exside opened the pull request:
It seemed not to be possible to get a list of all files in a folder represented by a modDirectory object, this method adds this functionality. It can do several things:

1) Scan the modDirectory path recursively or normally (non-recursive) with folders ignored (only filepaths are returned)
2) It can return or ignore (default) hidden (unix filesystems) files and folders starting with a dot (e.g. .htaccess etc.)
3) It can filter the returned filelist by file extensions passed via $options
4) It can sort the resulting filelist array alphabetically if desired (this maybe should be done by the user him/herself when working with the returned filelist array...but I thought it's maybe handy)

I think this could may be also of use for file mediasources, as they may implement a directoryiterator themselves, like this they could just make use of this method already provided by the filehandler...

usage example:

```
// get the filehandler service
$modx->getService('file','modFileHandler');

// create modFileHandler object, in this case we need a modDirectory object, eg. pass a folder path
$fileobj = $modx->file->make(MODX_ASSETS_PATH . 'site/test/');

// this most simple usage of this method just returns ALL files inside the $fileobj / modDirectory path non-recursively, unfiltered and unsorted BUT ignoring hidden files
$files = $fileobj->getFiles();

// now we want all files contained inside this folder, recursively but unsorted, filtered by css file extension
$files = $fileobj->getFiles(array('recursive' => true, 'sort' => false, 'extensions' => 'css', 'ignorehidden' => true));

// if we would like to sort the resulting filelist, the flags of array_multisort() can be passed, e.g. SORT_ASC, SORT_DESC, SORT_NUMERIC, SORT_NATURAL, SORT_REGULAR, note: they have to be passed WITHOUT quotes, e.g.
$files = $fileobj->getFiles(array('recursive' => true, 'sort' => SORT_ASC, 'extensions' => 'css', 'ignorehidden' => true));

echo '<pre>' . print_r($files,1) . '</pre>';
```

which then gives something like that:
![bildschirmfoto 2014-04-19 um 20 19 49](https://cloud.githubusercontent.com/assets/445501/2749160/46fcd0c2-c7ef-11e3-8206-6cf1391881c0.png)

with sort flag = true:
![bildschirmfoto 2014-04-19 um 20 21 10](https://cloud.githubusercontent.com/assets/445501/2749162/70f67f7c-c7ef-11e3-9564-8b6298d2107b.png)

with recursive flag = false, sort = true:
![bildschirmfoto 2014-04-19 um 20 22 02](https://cloud.githubusercontent.com/assets/445501/2749164/94fd77ae-c7ef-11e3-81ae-4c3a34d787e9.png)

you get the picture I hope =)

Commit:
Merge pull request #1 from modxcms/develop

Update from original

exside committed:
modFileHandler enhancement: added modDirectory getFiles() method

exside committed:
Improving the method based on feedback and make it compatible with PHP <= 5.3

exside committed:
Fix docblock description

Mark-H commented on core/model/modx/modfilehandler.class.php:
```diff
@@ -514,6 +514,70 @@ public function create($mode = '') {
+            foreach (new RecursiveIteratorIterator($iterator) as $file) {
+                $dirname = explode(DIRECTORY_SEPARATOR, $file->getPath()); // PHP Strict warning if we put end() around here
+                
+                if (($options['ignorehidden'] ? (substr($file->getFilename(), 0, 1) !== '.' && substr(end($dirname), 0, 1) !== '.') : true) && $file->isFile()) {
```
Hey Luk! I think this PR is great and _very_ useful, but this line (also repeated on 559 below) is really confusing and hard to read. In the spirit of leaving code better than before, any chance this could be made more readable? ;)

OptimusCrime replied:
I don't understand what is sooo confusing with those lines? He is just using the Iterator class.

Mark-H replied:
I meant the ternary inside an if with another condition at the end. Iterators are fine, just this specific line (545) seems to be choosing conciseness over readability. As I said, leaving things better than it was found ;)

exside replied:
=)...lol, definitely "choosing conciseness over readability" here...but how to avoid quadruplicating the following:

```php
           if (!empty($options['extensions'])) {
                            if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                                $files[] = $file->getPathname();
                            }
                        } else {
                            $files[] = $file->getPathname();
                        }
```
because if I take out the ugly line that checks everything then I have to wrap if / else statements around, which could look something like that, that is also not very pretty to me:

```php
                if ($options['ignorehidden'] && $file->isFile()) {

                    if (substr($file->getFilename(), 0, 1) !== '.' && substr(end($dirname), 0, 1) !== '.') {
                        if (!empty($options['extensions'])) {
                            if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                                $files[] = $file->getPathname();
                            }
                        } else {
                            $files[] = $file->getPathname();
                        }
                    }
                } else {
                    if (!empty($options['extensions'])) {
                        if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                            $files[] = $file->getPathname();
                        }
                    } else {
                        $files[] = $file->getPathname();
                    }                    
                }

```
did you mean something like that @Mark-H 

another option of keeping the tenary and avoid the code-quadruplication (it's already duplicate, that's ugly enough =D) would be to "extract" the long/ugly part into a variable, which would make the tenary a bit less ugly, e.g. the full code would look like that

```php
    public function getFiles($options = array()) {
        $options = array_merge(array(
            'recursive' => false,
            'ignorehidden' => true,
            'sort' => false,
            'extensions' => ''
        ), $options);
        $files = array();
        $extensions = explode(',', $options['extensions']);
        
        if ($options['recursive']) {
            $iterator = new RecursiveDirectoryIterator($this->path);
            foreach (new RecursiveIteratorIterator($iterator) as $file) {
                $dirname = explode(DIRECTORY_SEPARATOR, $file->getPath()); // PHP Strict warning if we put end() around here
                // would return true if first char in filename or dirname is a dot (.)
                $ishidden = substr($file->getFilename(), 0, 1) === '.' && substr(end($dirname), 0, 1) === '.';

                if (($options['ignorehidden'] ? !$ishidden : true) && $file->isFile()) {
                    if (!empty($options['extensions'])) {
                        if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                            $files[] = $file->getPathname();
                        }
                    } else {
                        $files[] = $file->getPathname();
                    }
                }
            }
        } else {
            $iterator = new DirectoryIterator($this->path);
            foreach ($iterator as $file) {
                // would return true if first char in filename is a dot (.)
                $ishidden = substr($file->getFilename(), 0, 1) === '.';

                if (($options['ignorehidden'] ? !$ishidden : true) && $file->isFile()) {
                    if (!empty($options['extensions'])) {
                        if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                            $files[] = $file->getPathname();
                        }
                    } else {
                        if ($file->isFile()) {
                            $files[] = $file->getPathname();
                        }
                    }
                }
            }
        }
        if (!empty($options['sort'])) {
            array_multisort($files, $options['sort'], $files);
        }
        return $files;
    }
```

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

opengeek committed:
xPDO 2.4.0-pl

opengeek committed:
xPDO 2.4.0-pl

Merge remote-tracking branch 'xpdo/master'

* xpdo/master:
  xPDO 2.4.0-pl

ExcaliburKG committed:
fixed "gzopen not found" issue for systems with this function named as "gzopen64"

rtripault committed:
Added modParsedManagerController (base manager controller class) which allows building CMPs making use of MODX tags (snippets, chunks...)

rtripault committed:
Allow Smarty template to have MODX tags too

rtripault committed:
Converted "help" manager page to use newly added modParsedManagerController

bezumkin committed:
Added check of "modx-resource-id" in treedrop

In MODX 2.4.0-pl drag-n-drop elements to template content is broken because of usage on undefined variable on line 227.
[![](https://file.modx.pro/files/0/8/d/08d37129cf625cee84d68dddc2d4127as.jpg)](https://file.modx.pro/files/0/8/d/08d37129cf625cee84d68dddc2d4127a.png)

This fix add check of "modx-resource-id" availability

zaigham committed:
Prints out the name of failing plugin in error log

## Before

[2015-08-21 21:46:46] (ERROR @ /connectors/index.php) [OnDocFormSave] Plugin failed!

## After

[2015-08-21 21:46:46] (ERROR @ /connectors/index.php) [OnDocFormSave] Plugin some-plugin-name-here failed!

krismas committed:
Really clear the log...

Remove the extra space when clearing « error.log » file via
`writeFile()`

zaigham committed:
Resolves #12578

jpdevries committed:
Updating Font-Awesome to 4.4

66 New Icons in 4.4
we don’t need a bower version and it is ok for the package to be private

Jako committed:
Wrong language topic

There exists no language topic 'sources' but 'source'

Jako committed:
Wrong language topic

There exists no language topic 'sources' but 'source'

Jako committed:
Added missing lexicon keys

Could be filled with real text and translated afterwards (has to be added to Crowdin then).

Jako committed:
Added missing lexicon keys

The file types are used in default thumbnailType options.

Fi1osof committed:
Bugfix for #12580

https://github.com/modxcms/revolution/issues/12580

opengeek committed:
Add anonymous_sessions to allow session-less access for anonymous users

Default behavior is unchanged and set to true. When disabled however, anonymous users will not start a session when accessing a page on the site; only logged in users will have access to a session.

Also adds modX->startSession() which starts a session if not already and simply returns true if a session is available, regardless how it was started

matdave committed:
Added patch for ExtJS Drag & Drop issue #12576

Added code provided by berniesaurus to fix issue where you couldn't drag and drop into a window or closable tab panel after one had already been rendered. https://www.sencha.com/forum/showthread.php?264400-Ext-JS-3.4.4.1-Drag-amp-Drop-broken

opengeek committed:
Correct API docs for modPackageBuilder->createVehicle()

Does not accept only xPDOObject instances; vehicles can be constructed from just about any payload.

Mark-H committed:
Fix loading rich text editors on non-document resource types

Mark-H committed:
Restore backwards compatibility for addons interacting with modTransportProvider->getClient

The getClient method was previously public (see https://github.com/modxcms/revolution/blob/v2.3.6-pl/core/model/modx/transport/modtransportprovider.class.php#L73) and used by Gitify to provide a way of installing packages via the command line.

Jako committed:
Not rendered output properties of custom TVs

### What does it do ?
Fixing not rendered output properties of custom TVs.

### Why is it needed ?
Not rendered output properties of custom TVs.

### Related issue(s)/PR(s)
See: https://github.com/Jako/ImagePlus/issues/13

Commit:
Fixed PHP version check to latest 5.3.3 requirement. Refactored code. Adjusted error messages.

rtripault committed:
Merge remote-tracking branch 'upstream/pr/12631' into pr/12631

* upstream/pr/12631:
  Correct API docs for modPackageBuilder->createVehicle()

rtripault committed:
Merge remote-tracking branch 'upstream/pr/12573' into pr/12573

* upstream/pr/12573:
  Prints out the name of failing plugin in error log

pixelchutes committed:
Update PHPMailer to v5.2.13

Resolves #12650

pixelchutes committed:
Convert method name references to camelCase

hansek committed:
Syntax cleanup

hansek committed:
Fix permissions for child policies if their template lost some

E.g.
If you remove "about" permission from AdministratorTemplate
Administrator policy will still have active "about" permission
which is wrong.

Solution
Loop over all Policies using same Policy template and remove those permissions.

rexterkate-adwise committed:
xpdo::getOption performance improvement

rexterkate-adwise committed:
xpdo::getOption performance improvement (v2)

rexterkate-adwise committed:
xpdo::getOption return default if $key isn't a string or an array

rexterkate-adwise committed:
xpdo::getOption fix skipEmpty logic

opengeek committed:
Optimize xPDO::getOption() for significant performance increase

Merge branch 'pr/74' into 2.x

opengeek committed:
Merge remote-tracking branch 'xpdo/2.4.x'

* xpdo/2.4.x:
  xpdo::getOption fix skipEmpty logic
  xpdo::getOption return default if $key isn't a string or an array
  xpdo::getOption performance improvement (v2)
  xpdo::getOption performance improvement

rexterkate-adwise committed:
xpdo::getOption fix skipEmpty logic (v2)

rexterkate-adwise committed:
xpdo::getOption fix skipEmpty logic (v3: performance tweak)

pixelchutes committed:
Restore PHPMailer autoloader

theboxer committed:
Restore backwards compatibility for addons interacting with modTransportProvider

* origin/pr/12633:
  Restore backwards compatibility for addons interacting with modTransportProvider->getClient

theboxer committed:
On policy template update sync policies with policy template

* origin/pr/12654:
  Fix permissions for child policies if their template lost some
  Syntax cleanup

theboxer committed:
Fix dropping elements in template

* origin/pr/12572:
  Added check of "modx-resource-id" in treedrop

opengeek committed:
Fix skipEmpty logic in xPDO::getOption() from #74

Merge remote-tracking branch 'origin/pr/75' into 2.4.x

* origin/pr/75:
  xpdo::getOption fix skipEmpty logic (v3: performance tweak)
  xpdo::getOption fix skipEmpty logic (v2)

theboxer committed:
Change user's grid in ACL view to be same as in update group. Allowing user manipulation in group.

opengeek committed:
xPDO 2.4.1-pl

opengeek committed:
Merge remote-tracking branch 'xpdo/2.4.x'

* xpdo/2.4.x:
  xPDO 2.4.1-pl
  fixed "gzopen not found" issue for systems with this function named as "gzopen64"
  xpdo::getOption fix skipEmpty logic (v3: performance tweak)
  xpdo::getOption fix skipEmpty logic (v2)

opengeek committed:
Update to xPDO 2.4.1-pl

Merge remote-tracking branch 'xpdo/master' into xpdo-2.4.1

* xpdo/master:
  xPDO 2.4.1-pl
  fixed "gzopen not found" issue for systems with this function named as "gzopen64"
  xpdo::getOption fix skipEmpty logic (v3: performance tweak)
  xpdo::getOption fix skipEmpty logic (v2)
  xpdo::getOption fix skipEmpty logic
  xpdo::getOption return default if $key isn't a string or an array
  xpdo::getOption performance improvement (v2)
  xpdo::getOption performance improvement
  xPDO 2.4.0-pl

theboxer committed:
Update xPDO to 2.4.1-pl

theboxer committed:
Fix panel height

theboxer committed:
Shorten adding listener to grid's store

theboxer committed:
Fix height

opengeek committed:
Make user grid in ACL view consistent with user group view

Merge remote-tracking branch 'origin/pr/12662' into 2.4.x

* origin/pr/12662:
  Fix height
  Shorten adding listener to grid's store
  Fix panel height
  Change user's grid in ACL view to be same as in update group. Allowing user manipulation in group.

Commit:
Update _tree.scss

opengeek committed:
Update PHPMailer to v5.2.13

Merge remote-tracking branch 'origin/pr/12652' into 2.4.x

* origin/pr/12652:
  Restore PHPMailer autoloader
  Convert method name references to camelCase
  Update PHPMailer to v5.2.13

opengeek committed:
MODX Revolution 2.4.1-pl

tyllo committed:
bugfix #131 in modmore/Gitify

Fix issue: https://github.com/modmore/Gitify/issues/131

tyllo committed:
Merge pull request #1 from tyllo/fix#Gitify#131

bugfix #131 in modmore/Gitify

Jako committed:
Fix image width and add transparency pattern

Jako committed:
Fix image width and add transparency pattern

sergant210 committed:
Fixed emptyRecycleBin action.

In the resource tree panel, after empty Trash action the icon is still available and shows wrong description.

sergant210 committed:
Fire  'emptyTrash'  event on emptyRecycleBin

### What does it do ?
Fire js 'emptyTrash'  event on emptying recycle bin action like event 'OnEmptyTrash' for MODX plugins.

### Why is it needed ?
For interact with thirdparty Extra.

argnist committed:
Prevent id of property set to be text

The id should be 0 if default properties are chosen. The value 0 properly shows "Default" text so we have no visible difference.

ilyautkin committed:
Fixed group settings updatins

In parent::initialize() object will getting again, but by primaryKeyField

Omeryl committed:
Add slackin badge to README.md

Adds a slackin badge, as well as a link to the http://modx.org/ site for the MODX Community slack. Implements #12636.

Omeryl committed:
Update Slackin badge in README.md to use HTTPS

theboxer committed:
Add slackin badge to README.md

* origin/pr/12680:
  Update Slackin badge in README.md to use HTTPS
  Add slackin badge to README.md

theboxer committed:
MAdd slackin badge to README.md

* origin/pr/12680:
  Update Slackin badge in README.md to use HTTPS
  Add slackin badge to README.md

theboxer committed:
Pass 0 as id of default property set instead of "Default"

* origin/pr/12674:
  Prevent id of property set to be text

theboxer committed:
Disable trash icon and set proper tooltip after removing resources

* origin/pr/12672:
  Fixed emptyRecycleBin action.

theboxer committed:
Change version to 2.4.2-dev

theboxer committed:
Fix image width and add transparency pattern

* origin/pr/12670:
  Fix image width and add transparency pattern
  Fix image width and add transparency pattern

theboxer committed:
Fix not rendering output properties of custom TVs

* origin/pr/12635:
  Not rendered output properties of custom TVs

theboxer committed:
Merge remote-tracking branch 'origin/pr/12602' into 2.4.x

* origin/pr/12602:
  Added missing lexicon keys
  Added missing lexicon keys
  Wrong language topic
  Wrong language topic

theboxer committed:
Add hover effect to tree expand/collapse icon

* origin/pr/12664:
  Update _tree.scss

theboxer committed:
Update setup to check the minimum supported PHP version

* origin/pr/12638:
  Fixed PHP version check to latest 5.3.3 requirement. Refactored code. Adjusted error messages.

theboxer committed:
Merge remote-tracking branch 'origin/pr/12605' into 2.4.x

* origin/pr/12605:
  Really clear the log...

theboxer committed:
Update Font-Awesome to 4.4

* origin/pr/12598:
  Updating Font-Awesome to 4.4

theboxer committed:
Add resource title in Manager Log for edited resources

* origin/pr/12589:
  Resolves #12578

theboxer committed:
Merge branch '2.4.x' into 2.x

* 2.4.x:
  Change version to 2.4.2-dev
  Prevent id of property set to be text
  Fixed emptyRecycleBin action.
  Fix image width and add transparency pattern
  Fix image width and add transparency pattern
  Update _tree.scss
  Fixed PHP version check to latest 5.3.3 requirement. Refactored code. Adjusted error messages.
  Not rendered output properties of custom TVs
  Added missing lexicon keys
  Added missing lexicon keys
  Wrong language topic
  Wrong language topic
  Updating Font-Awesome to 4.4
  Resolves #12578
  Really clear the log...

alroniks committed:
Added missed target=blank

theboxer committed:
Merge remote-tracking branch 'origin/pr/12686' into 2.4.x

* origin/pr/12686:
  Added missed target=blank

theboxer committed:
Fix initialization of modUserGroupSettingUpdateProcessor processor

* origin/pr/12678:
  Fixed group settings updatins

theboxer committed:
Add patch for ExtJS Drag & Drop issue

* origin/pr/12617:
  Added patch for ExtJS Drag & Drop issue #12576

Commit:
Fixed tree style bug 12644.
Fixed typo in background color.

Commit:
Fixed tree style bug 12644.
Fixed typo in background color.

Commit:
Implements #12690 by adding a unique token to manager js and css links to enable automatic refresh after upgrade.

theboxer committed:
Different tree styles for unpublished + hidemenu

* origin/pr/12699:
  Fixed tree style bug 12644. Fixed typo in background color.
  Fixed tree style bug 12644. Fixed typo in background color.

theboxer committed:
Merge remote-tracking branch 'origin/pr/12614' into property-sets

* origin/pr/12614:
  Bugfix for #12580

theboxer committed:
Updates for the modPropertySetUpdateFromElementProcessor processor

theboxer committed:
Fix emptying property sets on element save

* property-sets:
  Updates for the modPropertySetUpdateFromElementProcessor processor
  Bugfix for #12580

theboxer committed:
Revert "Set default background color for thumbs to WHITE instead of BLACK"

This reverts commit 8896c78a040e190e0548db7d9670aa771ef9a2ab.

theboxer committed:
Update lexicons from crowdin

Crowdin project: http://translate.modx.com
Thanks to all translators and proofreaders for contribution!

theboxer committed:
MODX Revolution 2.4.2-pl

Commit:
Added examples to TV option "Where Condition" to simplify usage.

hansek committed:
Add option to allow define custom redirect method for each action button

Jako committed:
Bugfix for failure message

$f was removed in https://github.com/modxcms/revolution/commit/9a46c45aaf1879893d0cf162230dcdae99d94902

Jako committed:
Missing semicolon

Jako committed:
Better variable names

oori committed:
Corrected dashboard's save button action

Commit:
Fixed #12710 tree style for deactivated plugins.

theboxer committed:
Change version to 2.4.3-dev

theboxer committed:
Fix tree style for deactivated plugin

* origin/pr/12712:
  Fixed #12710 tree style for deactivated plugins.

theboxer committed:
Fix failure message in Export processor

* origin/pr/12709:
  Better variable names
  Missing semicolon
  Bugfix for failure message

theboxer committed:
Fix dashboard and dashboard widget save button

* origin/pr/12711:
  Corrected dashboard's save button action

theboxer committed:
Fix loading rich text editors on non-document resource types

* origin/pr/12632:
  Fix loading rich text editors on non-document resource types

theboxer committed:
Add examples to TV option "Where Condition" to simplify usage.

* origin/pr/12702:
  Added examples to TV option "Where Condition" to simplify usage.

theboxer committed:
Change version to 2.5.0-dev

theboxer committed:
Add anonymous_sessions to allow session-less access for anonymous users

* origin/pr/12616:
  Add anonymous_sessions to allow session-less access for anonymous users

theboxer committed:
Fire emptyTrash event after emptying recycle bin

* origin/pr/12673:
  Fire  'emptyTrash'  event on emptyRecycleBin

theboxer committed:
Allow custom redirect method for each action button in action bar

* origin/pr/12703:
  Add option to allow define custom redirect method for each action button

hansek committed:
Fix Profile  Password reset to have same field labels as User panel password reset

Thanks to Dan Roucka for this issue ;)

Jako committed:
Fixing #12735 missing date.timezone setting on install

Jako committed:
Fixing #12735 missing date.timezone setting on install

Jako committed:
Merge branch '2.x' of https://github.com/Jako/revolution into 2.x

rtripault committed:
Pass modX object by reference as expected

rtripault committed:
Try to make use of the maximum available height for logs textarea

Commit:
Added resource ids to OnResourceAutoPublish param

rtripault committed:
Added some more flexibility is search results labels/icons (mostly in case of overrides)

Commit:
500 error editing user which doesn't exist

Fixing PHP Fatal error when trying to edit a user which doesn't exist. The page title references the user, which doesn't exist.

exside committed:
Improve usability of tree by trying to prevent accidential editing of resources when container should be toggled

exside committed:
Add new class based processor for unpacking archives

exside committed:
Add new permission file_unpack with corresponding lexicon string

exside committed:
Add unpack() method to modFileHandler

exside committed:
Commit functional code for unpacking files via FileTree

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12555' into 2.x

* upstream/pr/12555:
  Converted "help" manager page to use newly added modParsedManagerController
  Allow Smarty template to have MODX tags too
  Added modParsedManagerController (base manager controller class) which allows building CMPs making use of MODX tags (snippets, chunks...)

jpdevries committed:
Making header more mobile friendly

Gets the fixed and absolute positioned silliness out of the way on
small screens

jpdevries committed:
Making lefthand sidebar (tree) more mobile friendly

jpdevries committed:
Make Manager Edit page more mobile friendly

woohoo
Content tab
Settings tab

jpdevries committed:
Responsive collapse / expand sidebar

It’s a little weird because the component is all the way at the bottom
of the page. Ideally it would be just beneath the tree sidebar itself,
but we haven’t been able to figure out how to change where the
component is ordered within the markup so this is the best we can do
for now

exside committed:
Allow Uberbar search for resource ID...

jpdevries committed:
Updating font-awesome to ^4.5.0

Closes modxcms#12770

jpdevries committed:
more mobile enhancements (momo)

Import HTML
Import Static HTML
Resource Groups
Content Types
Media Sources
Installer
Manager Customizations
Dashboard
Contexts
Menus
ACLs
Property Sets
Lexicon Management
Namespace

exside committed:
Show resource ID in search results when permission allows it

exside committed:
Add move() method to filehandler, alias for rename()

jpdevries committed:
Fix wrong FontAwesome icon "icon-coffee"

closes modxcms/revolution#12676

jpdevries committed:
html-first autofocus on login

currently the username receives focus via js, this does the same with
html5

pixelchutes committed:
Allow searching by Resource ID in Uberbar

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12776' into 2.x

* upstream/pr/12776:
  more mobile enhancements (momo)
  Responsive collapse / expand sidebar
  Make Manager Edit page more mobile friendly
  Making lefthand sidebar (tree) more mobile friendly
  Making header more mobile friendly

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12784' into 2.x

* upstream/pr/12784:
  Fix wrong FontAwesome icon "icon-coffee"

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12774' into 2.x

* upstream/pr/12774:
  Updating font-awesome to ^4.5.0

Conflicts:
	_build/templates/default/bower.json

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12746' into 2.x

* upstream/pr/12746:
  Try to make use of the maximum available height for logs textarea

jpdevries committed:
Improve keyboard / screen reader a11y of login screen

Tab order is now:
username
password
forgot your login
remember me
login
MODX link
language select

Closes modxcms/revolution#12145 (for reals)
Closes modxcms/revolution#12146 (for reals)
Closes modxcms/revolution#12147 (for reals)

Also see modxcms/revolution#12782
Also see modxcms/revolution#12791

Adds both aria-required="true" required and required attributes
although i think only required may be needed

jpdevries committed:
Fix uberbar on desktop

Closes modxcms/revolution#12792

opengeek committed:
Change modResource Children to Composite relation

This was redone via the build script and applied to sqlsrv as well
Closes #12279

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12795' into 2.x

* upstream/pr/12795:
  Fix uberbar on desktop

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12773' into 2.x

* upstream/pr/12773:
  Improve usability of tree by trying to prevent accidential editing of resources when container should be toggled

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12747' into 2.x

* upstream/pr/12747:
  Added resource ids to OnResourceAutoPublish param

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12772' into 2.4.x

* upstream/pr/12772:
  500 error editing user which doesn't exist

Mark-H committed:
Fix CURLOPT_FOLLOWLOCATION warning in logs

CURLOPT_FOLLOWLOCATION not works with `safe_mode` or `open_basedir` enabled. So config_check processor generates such entries in error log:

> [2015-10-31 09:00:50] (ERROR @ .../www/core/model/modx/processors/system/config_check.inc.php : 97) PHP warning: curl_setopt(): CURLOPT_FOLLOWLOCATION cannot be activated when an open_basedir is set

This will fix it.
(cherry picked from commit 2ab0560588542ded045d9a8538a978944e1f932e)

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12700' into 2.4.x

* upstream/pr/12700:
  Implements #12690 by adding a unique token to manager js and css links to enable automatic refresh after upgrade.

Mark-H committed:
Ensure compatibility of more mobile-friendly manager with custom manager themes

The more mobile-friendly manager introduced in #12776 relies on a new div with id modx-action-buttons-container being present in the header.tpl, however with custom manager themes that have their own header.tpl file that may not always be the case. With this commit in place, it ensures the action buttons are still inserted into the page (albeit in the wrong page), rather than throwing a js error if it can't find the new div.

Mark-H committed:
Fix scrolling on non-mobile displays

This bug was introduced in https://github.com/modxcms/revolution/commit/b9d4e741278c024243fb288218ed12ec6502ceb7 and caused scrolling the page to be broken on large displays (whoops). Limiting the auto width/height to mobile screens fixes that.

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12741' into 2.4.x

* upstream/pr/12741:
  Pass modX object by reference as expected

pixelchutes committed:
Shows a 'please wait' message while modSystemRefreshurisProcessor is running.

rtripault committed:
Merge remote-tracking branch 'upstream/pr/12799' into 2.x

* upstream/pr/12799:
  Fix scrolling on non-mobile displays

rtripault committed:
Merge branch '2.4.x' into 2.x

* 2.4.x:
  Fix CURLOPT_FOLLOWLOCATION warning in logs
  500 error editing user which doesn't exist
  Pass modX object by reference as expected
  Implements #12690 by adding a unique token to manager js and css links to enable automatic refresh after upgrade.

rtripault committed:
Merge remote-tracking branch 'upstream/pr/12798' into 2.x

* upstream/pr/12798:
  Ensure compatibility of more mobile-friendly manager with custom manager themes

rtripault committed:
Added PHP7 & HHVM to be checked against

rtripault committed:
Allow HHVM failure

rtripault committed:
Merge remote-tracking branch 'upstream/pr/12721' into 2.4.x

* upstream/pr/12721:
  Fix Profile  Password reset to have same field labels as User panel password reset

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12801' into 2.x

* upstream/pr/12801:
  Allow HHVM failure
  Added PHP7 & HHVM to be checked against

Mark-H committed:
Fix branch for travis badge in readme

pixelchutes committed:
Merge pull request #12803 from Mark-H/fix-travis-branch

Fix branch for travis badge in readme

Mark-H committed:
Fix travis builds on PHP 5.5

Based on fix in http://andreas.heigl.org/2015/11/12/testing-code-with-phpunit-on-travis-ci-for-php-5-5-and-php7/ and travis-ci/travis-ci#5206, this adjustment makes sure that on PHP 5.5, phpunit 4.x is loaded rather than 5.x which is not compatible with PHP 5.5, but required for PHP7.

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12794' into 2.x

* upstream/pr/12794:
  Improve keyboard / screen reader a11y of login screen

Mark-H committed:
Merge branch '2.x' of github.com:modxcms/revolution into 2.x

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12787' into 2.x

* upstream/pr/12787:
  html-first autofocus on login

Conflicts:
	manager/templates/default/security/login.tpl

pixelchutes committed:
Merge pull request #12804 from Mark-H/fix-travis-build-on-php55

Fix travis builds on PHP 5.5

exside committed:
Replace smarty with version 3.1.27

exside committed:
Update phpmailer to version 5.2.14

exside committed:
Replace use of removed ereg() function in modphpthumb.class.php

exside committed:
Update system info

exside committed:
Update system info

exside committed:
Fix more occurences of ereg/eregi

exside committed:
Fix preg_match(): Unknown modifier 'h'

exside commented:
@Mark-H any thoughts on the last comment?

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12808' into 2.x

* upstream/pr/12808:
  Update system info
  Update phpmailer to version 5.2.14

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12807' into 2.x

* upstream/pr/12807:
  Update system info
  Replace smarty with version 3.1.27

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12749' into 2.x

* upstream/pr/12749:
  Added some more flexibility is search results labels/icons (mostly in case of overrides)

pixelchutes committed:
Sets correct responseType in the rest service for multiple packages via a single request

Fixes modmore/Gitify#131

Commit:
added unpub_date to publishingResults

Extend $publishingResults with `unpub_date` which is returned during the autoPublish event.

Mark-H committed:
Update login.class.php

fix the rememberme unsaved problem
(cherry picked from commit 04dcb0993daa3810009ddbbaf0676925f5f49a5d)

Mark-H committed:
Merge branch '2.4.x' of github.com:modxcms/revolution into 2.4.x

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12800' into 2.4.x

* upstream/pr/12800:
  Shows a 'please wait' message while modSystemRefreshurisProcessor is running.

Mark-H committed:
Merge branch '2.4.x' into 2.x

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12775' into 2.x

* upstream/pr/12775:
  Add move() method to filehandler, alias for rename()
  Commit functional code for unpacking files via FileTree
  Add unpack() method to modFileHandler
  Add new permission file_unpack with corresponding lexicon string
  Add new class based processor for unpacking archives

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12814' into 2.x

* upstream/pr/12814:
  added unpub_date to publishingResults

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12738' into 2.x

* upstream/pr/12738:
  Fixing #12735 missing date.timezone setting on install
  Fixing #12735 missing date.timezone setting on install

Mark-H committed:
Merge remote-tracking branch 'upstream/pr/12809' into 2.x

* upstream/pr/12809:
  Fix preg_match(): Unknown modifier 'h'
  Fix more occurences of ereg/eregi
  Replace use of removed ereg() function in modphpthumb.class.php

exside committed:
Rebase on 2.x and refactor method

exside committed:
Make skiphidden option more reliable

exside committed:
Add ability to skip individual files

exside committed:
Add ability to use a callback function to modify the output on the fly

exside commented:
Alright, took a second jump at this thing and improved it a bit more (hopefully). The whole method now looks like this:

```php
    /**
     * Iterates over a modDirectory object and returns an array of all containing files,
     * optionally runs recursive, filters by file extension(s) and sorts the resulting list according to a specified sort flag
     *
     * @param array $options Options for iterating the directory.
     * @option boolean recursive If also subfolders should be scanned for files
     * @option boolean|string sort If the resulting filelist array should be sorted with the specified flag like SORT_ASC, SORT_DESC, SORT_REGULAR, SORT_NATURAL, SORT_NUMERIC
     * @option boolean skiphidden If folders and files starting with a dot . (hidden files/folders in unix envirionments) should be ignored, defaults to true
     * @option string|array skip Comma separated list or array of filenames (including extension) that should be ignored
     * @option string|array extensions Comma separated list or array of file extensions to filter files by
     * @option boolean|function callback Anonymous function to modify each output item, $file will be passed / accessible
     *      
     * @return array
     */
    public function getFiles($options = array()) {
        $options = array_merge(array(
            'recursive' => false,
            'sort' => false,
            'skiphidden' => true,
            'skip' => array(),
            'extensions' => array(),
            'callback' => false,
        ), $options);

        $files = array();
        $extensions = !is_array($options['extensions']) ? explode(',', $options['extensions']) : $options['extensions'];
        $skip = !is_array($options['skip']) ? explode(',', $options['skip']) : $options['skip'];
        $iterator = $options['recursive'] ? new RecursiveIteratorIterator(new RecursiveDirectoryIterator($this->path)) : new DirectoryIterator($this->path);

        foreach ($iterator as $file) {
            if ($options['skiphidden']) {
                // check for hidden folder, also hide with visible ones inside
                // but don't skip weird filenames like "...and-there-was-silence.avi"
                if (preg_match('/(\/\.\w+|\\\.\w+)/', $file->getPath())) {
                    continue;
                }
                // check for hidden file (probably works only on UNIX filesystems)
                $ishidden = preg_match('/^(\.\w+)/i', $file->getFilename());
            } else {
                $ishidden = false;
            }

            if ($file->isFile() && ($options['skiphidden'] ? !$ishidden : true) && (!empty($skip) ? !in_array($file->getFilename(), $skip) : true)) {
                $addfile = false;
                
                if (!empty($options['extensions'])) {
                    if (in_array(pathinfo($file->getPathname(), PATHINFO_EXTENSION), $extensions)) {
                        $addfile = true;
                    }
                } else {
                    $addfile = true;
                }

                if ($addfile) {
                    if (is_object($options['callback']) && ($options['callback'] instanceof Closure)) {
                        $files[] = $options['callback']($file);
                    } else {
                        $files[] = $file->getPathname();
                    }
                }
            }
        }

        if (!empty($options['sort'])) {
            array_multisort($files, $options['sort'], $files);
        }

        return $files;
    }
```

The method is now a bit shorter (not so much code duplication) and much more powerful and hopefully also a bit more readable (what do you think @Mark-H)! As MODX now has a minimal required PHP version of 5.3, we can use anonymous functions, which allow for a "callback" functionality, which puts this method on a whole new level...the normal behaviour is still like explained in the initial message, only skipping hidden files and folders was improved a bit (it listed files in visible folders inside hidden folders... e.g. /somedir/.somehiddendir/visibledir/file.txt) and now it's also possible to filter out very specific files by filename (not just inclusion/filtering by extension).

An advanced call could look like this:

```php
$files = $fileobj->getFiles(array(
	'recursive' => true,
	'sort' => false,
	'extensions' => '',
	'skiphidden' => true,
	'skip' => 'image_a.png,Image_A.png',
	'callback' => function($file) { 
		return array(
		'user' => $this->modx->user->id,
		'filename' => $file->getFilename(),
		'path' => $file->getPathname(),
		'perms' => $file->getPerms());
	}
));

// which would result in something like that:

Array
(
    [0] => Array
        (
            [user] => 1
            [filename] => IMAGE_A.png
            [path] => /path/to/media/media.test/IMAGE_A.png
            [perms] => 33188
        )

    [1] => Array
        (
            [user] => 1
            [filename] => image_b.png
            [path] => /path/to/media/media.test/image_b.png
            [perms] => 33188
        )

    [2] => Array
        (
            [user] => 1
            [filename] => Image_B.png
            [path] => /path/to/media/media.test/Image_B.png
            [perms] => 33188
        )

    [3] => Array
        (
            [user] => 1
            [filename] => IMAGE_B.png
            [path] => /path/to/media/media.test/IMAGE_B.png
            [perms] => 33188
        )

    [4] => Array
        (
            [user] => 1
            [filename] => image_c.png
            [path] => /path/to/media/media.test/image_c.png
            [perms] => 33188
        )

    [5] => Array
        (
            [user] => 1
            [filename] => Image_C.png
            [path] => /path/to/media/media.test/Image_C.png
            [perms] => 33188
        )

    [6] => Array
        (
            [user] => 1
            [filename] => IMAGE_C.png
            [path] => /path/to/media/media.test/IMAGE_C.png
            [perms] => 33188
        )
)
```

as you can see we are now not anymore limited to "just" getting a list of filepaths, we can completely customize the output of the method with the callback function, that makes it very suitable as a replacement for directory listings in media source code, for example...or basically wherever the files of a directory need to be listed...

exside commented:
hups...was rebasing a bad idea?

exside committed:
Shorten/optimize if/else logic

exside committed:
Improve sort, enable multibyte, allow listing of folders, micro optimizations, rename method to list()

exside committed:
Rename method again, list is illegal...no getList()

theboxer commented:
@exside would you mind creating new PR? This one got a bit messy...

exside commented:
@theboxer sorry, was in holidays^^...looks like you managed to merge it anyways?
